addons/mass_mailing/models/mail_notification.py

```diff
@@ -1,41 +1,32 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
+from dateutil.relativedelta import relativedelta
+
 from odoo import api, fields, models
 
 
-class MailingTrace(models.Model):
-    """ MailingTrace models the statistics collected about emails. Those statistics
-    are stored in a separated model and table to avoid bloating the mail_mail table
-    with statistics values. This also allows to delete emails send with mass mailing
-    without loosing the statistics about them. """
-    _name = 'mailing.trace'
-    _description = 'Mailing Statistics'
-    _rec_name = 'id'
-    _order = 'scheduled DESC'
+class Notification(models.Model):
+    # Extend the mail.notification model for tracking (marketing purpose)
+    _inherit = 'mail.notification'
 
-    trace_type = fields.Selection([('email', 'Mail')], string='Type', default='email', required=True)
     display_name = fields.Char(compute='_compute_display_name')
     # mail data
-    mail_id = fields.Many2one('mail.mail', string='Mail', index=True)
     mail_id_int = fields.Integer(
         string='Mail ID (tech)',
         help='ID of the related mail_mail. This field is an integer field because '
              'the related mail_mail can be deleted separately from its statistics. '
              'However the ID is needed for several action and controllers.',
-        index=True,
-    )
+        index=True)
     email = fields.Char(string="Email")
     message_id = fields.Char(string='Message-ID', help="Technical field for the email Message-ID (RFC 2392)")
     # document
     model = fields.Char(string='Document model')
-    res_id = fields.Integer(string='Document ID')
+    res_id = fields.Many2oneReference(string='Document ID', model_field='model')
     # campaign / wave data
+    # if mass_mailing_id is set the notification will act as a trace
     mass_mailing_id = fields.Many2one('mailing.mailing', string='Mailing', index=True, ondelete='cascade')
-    campaign_id = fields.Many2one(
-        related='mass_mailing_id.campaign_id',
-        string='Campaign',
-        store=True, readonly=True, index=True)
+    campaign_id = fields.Many2one('utm.campaign', string='Campaign', store=True, readonly=False, index=True)
```

Comment:
`store=True, readonly=False` are the default values for a regular field, you can remove those.